ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,18 +702,91 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
+    // replica1 will start returning entries from the next blob which resides after logEndOffSetOnStartup.
+    // In other words, one blob will be missed out in findEntriesSince()
+    recovery = new MessageStoreRecovery() {
+      @Override
+      public List<MessageInfo> recover(Read read, long startOffset, long endOffset, StoreKeyFactory factory)
+          throws IOException {
+        return Collections.emptyList();
+      }
+    };
+    reloadIndex(true);
+    lastRecordOffset = index.journal.getLastOffset();
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on start up after restart.
+    startToken = new StoreFindToken(
+        new Offset(log.getEndOffset().getName(), (lastRecordOffset.getOffset() + 2 * PUT_RECORD_SIZE)), oldSessionId,
+        incarnationId);
+    // end token should point to offset of last entry i.e. firstRecordFileSpan.getEndOffset()
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId, incarnationId);
+    expectedEndToken.setBytesRead(bytesRead);
+    expectedEndToken.setInclusive();
+    // Fetch the FindToken returned from findEntriesSince
+    FindInfo findInfo = index.findEntriesSince(startToken, Long.MAX_VALUE);
+    Assert.assertEquals("EndToken mismatch ", expectedEndToken, findInfo.getFindToken());
+    Assert.assertEquals("No entries should have been returned ", 0, findInfo.getMessageEntries().size());
+
+    // add 2 new entries to the log
+    Offset endOffsetOfPrevMsg = index.getCurrentEndOffset();
+    appendToLog(PUT_RECORD_SIZE);
+    FileSpan fileSpan = log.getFileSpanForMessage(endOffsetOfPrevMsg, PUT_RECORD_SIZE);
+    IndexValue value = new IndexValue(PUT_RECORD_SIZE, fileSpan.getStartOffset(), Utils.Infinite_Time);
+    MockId id = getUniqueId();
+    index.addToIndex(new IndexEntry(id, value), fileSpan);
+    allKeys.put(id, new Pair<>(value, null));
+    endOffsetOfPrevMsg = fileSpan.getEndOffset();
+    appendToLog(PUT_RECORD_SIZE);
+
+    FileSpan fileSpan2 = log.getFileSpanForMessage(endOffsetOfPrevMsg, PUT_RECORD_SIZE);
+    IndexValue value2 = new IndexValue(PUT_RECORD_SIZE, fileSpan2.getStartOffset(), Utils.Infinite_Time);
+    MockId id2 = getUniqueId();
+    index.addToIndex(new IndexEntry(id2, value2), fileSpan2);
+    allKeys.put(id2, new Pair<>(value2, null));
+
+    Set<MockId> expectedEntries = new HashSet<>();
+    expectedEntries.add(id);
+    expectedEntries.add(id2);
+    bytesRead = log.getDifference(fileSpan2.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    expectedEndToken = new StoreFindToken(fileSpan2.getStartOffset(), sessionId, incarnationId);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest((StoreFindToken) findInfo.getFindToken(), Long.MAX_VALUE, expectedEntries, expectedEndToken);
```

Comment:
I might have old code - but where and how do you ensure that the returned entries contain exactly what is present in `expectedEntries`, since you are not passing in all of `findInfo`, but just the token?

Basically - have you verified that this test fails if the `inclusive` flag is not set in the token?